src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -85,172 +89,299 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
+                        index++;
+                        results.Add(resourceKey, new UpsertOutcome(resource, resource.Version == InitialVersion ? SaveOutcomeType.Created : SaveOutcomeType.Updated));
                     }
 
-                    _compressedRawResourceConverter.WriteCompressedRawResource(stream, resource.RawResource.Data);
-
-                    stream.Seek(0, 0);
+                    if (mergeWrappers.Count > 0) // do not call db with empty input
+                    {
+                        using var conn = await _sqlConnectionWrapperFactory.ObtainSqlConnectionWrapperAsync(cancellationToken, true); // TODO: Remove tran enlist when true bundle logic is in place.
+                        using var cmd = conn.CreateNonRetrySqlCommand();
+                        VLatest.MergeResources.PopulateCommand(
+                            cmd,
+                            AffectedRows: 0,
+                            RaiseExceptionOnConflict: true,
+                            IsResourceChangeCaptureEnabled: _coreFeatures.SupportsResourceChangeCapture,
+                            tableValuedParameters: _mergeResourcesTvpGeneratorVLatest.Generate(mergeWrappers));
+                        cmd.CommandTimeout = 180 + (int)(3600.0 / 20000 * results.Count);
+                        await cmd.ExecuteNonQueryAsync(cancellationToken);
+                    }
 
-                    _logger.LogInformation("Upserting {ResourceTypeName} with a stream length of {StreamLength}", resource.ResourceTypeName, stream.Length);
+                    await MergeResourcesCommitTransactionAsync(minSurrId, cancellationToken);
 
-                    // throwing ServiceUnavailableException in order to send a 503 error message to the client
-                    // indicating the server has a transient error, and the client can try again
-                    if (stream.Length < 31) // rather than error on a length of 0, a stream with a small number of bytes should still throw an error. RawResource.Data = null, still results in a stream of 29 bytes
+                    return results;
+                }
+                catch (SqlException e)
+                {
+                    // we cannot retry on connection loss as this call might be in outer transaction.
+                    // TODO: Add retries when set bundle processing is in place.
+                    if (e.Number == SqlErrorCodes.Conflict && retries++ < 10) // retries on conflict should never be more than 1, so it is OK to hardcode.
                     {
-                        _logger.LogCritical("Stream size for resource of type: {ResourceTypeName} is less than 50 bytes, request method: {RequestMethod}", resource.ResourceTypeName, resource.Request.Method);
-                        throw new ServiceUnavailableException();
+                        _logger.LogWarning(e, $"Error from SQL database on {nameof(MergeAsync)} retries={retries}");
```

Comment:
nit: Try not to use string interpolation for variables in log messages
```suggestion
                        _logger.LogWarning(e, $"Error from SQL database on {nameof(MergeAsync)} retries={{Retries}}", retries);
```

Reply:
will change